File: samples/csharp_dotnetcore/81.skills-skilldialog/DialogSkillBot/SkillAdapterWithErrorHandler.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Integration.AspNet.Core;
using Microsoft.Bot.Builder.TraceExtensions;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace Microsoft.BotBuilderSamples.DialogSkillBot
{
    public class SkillAdapterWithErrorHandler : BotFrameworkHttpAdapter
    {
        private readonly ConversationState _conversationState;
        private readonly ILogger _logger;

        public SkillAdapterWithErrorHandler(IConfiguration configuration, ICredentialProvider credentialProvider, AuthenticationConfiguration authConfig, ILogger<BotFrameworkHttpAdapter> logger, ConversationState conversationState)
            : base(configuration, credentialProvider, authConfig, logger: logger)
        {
            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
            OnTurnError = HandleTurnError;
        }

        private async Task HandleTurnError(ITurnContext turnContext, Exception exception)
        {
            // Log any leaked exception from the application.
            _logger.LogError(exception, $"[OnTurnError] unhandled error : {exception.Message}");

            await SendErrorMessageAsync(turnContext, exception);
            await SendEoCToParentAsync(turnContext, exception);
            await ClearConversationStateAsync(turnContext);
        }

        private async Task SendErrorMessageAsync(ITurnContext turnContext, Exception exception)
        {
            try
            {
                // Send a message to the user.
                var errorMessageText = "The skill encountered an error or bug.";
                var errorMessage = MessageFactory.Text(errorMessageText, errorMessageText, InputHints.IgnoringInput);
                await turnContext.SendActivityAsync(errorMessage);

                errorMessageText = "To continue to run this bot, please fix the bot source code.";
                errorMessage = MessageFactory.Text(errorMessageText, errorMessageText, InputHints.ExpectingInput);
                await turnContext.SendActivityAsync(errorMessage);

                // Send a trace activity, which will be displayed in the Bot Framework Emulator.
                // Note: we return the entire exception in the value property to help the developer;
                // this should not be done in production.
                await turnContext.TraceActivityAsync("OnTurnError Trace", exception.ToString(), "https://www.botframework.com/schemas/error", "TurnError");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Exception caught in SendErrorMessageAsync : {ex}");
            }
        }

        private async Task SendEoCToParentAsync(ITurnContext turnContext, Exception exception)
        {
            try
            {
                // Send an EndOfConversation activity to the skill caller with the error to end the conversation,
                // and let the caller decide what to do.
                var endOfConversation = Activity.CreateEndOfConversationActivity();
                endOfConversation.Code = "SkillError";
                endOfConversation.Text = exception.Message;
                await turnContext.SendActivityAsync(endOfConversation);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Exception caught in SendEoCToParentAsync : {ex}");
            }
        }

        private async Task ClearConversationStateAsync(ITurnContext turnContext)
        {
            try
            {
                // Delete the conversationState for the current conversation to prevent the
                // bot from getting stuck in a error-loop caused by being in a bad state.
                // ConversationState should be thought of as similar to "cookie-state" for a Web page.
                await _conversationState.DeleteAsync(turnContext);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Exception caught on attempting to Delete ConversationState : {ex}");
            }
        }
    }
}

File: samples/java_springboot/81.skills-skilldialog/dialog-skill-bot/src/main/java/com/microsoft/bot/sample/dialogskillbot/SkillAdapterWithErrorHandler.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MT License.

package com.microsoft.bot.sample.dialogskillbot;

import java.util.concurrent.CompletableFuture;

import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.builder.OnTurnErrorHandler;
import com.microsoft.bot.builder.TurnContext;
import com.microsoft.bot.connector.Async;
import com.microsoft.bot.connector.authentication.AuthenticationConfiguration;
import com.microsoft.bot.integration.BotFrameworkHttpAdapter;
import com.microsoft.bot.integration.Configuration;
import com.microsoft.bot.schema.Activity;
import com.microsoft.bot.schema.EndOfConversationCodes;
import com.microsoft.bot.schema.InputHints;

public class SkillAdapterWithErrorHandler extends BotFrameworkHttpAdapter {

    public SkillAdapterWithErrorHandler(
        Configuration configuration,
        AuthenticationConfiguration authenticationConfiguration
    ) {
        super(configuration, authenticationConfiguration);
        setOnTurnError(new SkillAdapterErrorHandler());
    }

    private class SkillAdapterErrorHandler implements OnTurnErrorHandler {

        @Override
        public CompletableFuture<Void> invoke(TurnContext turnContext, Throwable exception) {
            return sendErrorMessage(turnContext, exception).thenAccept(result -> {
                sendEoCToParent(turnContext, exception);
            });
        }

        private CompletableFuture<Void> sendErrorMessage(TurnContext turnContext, Throwable exception) {
            try {
                // Send a message to the user.
                String errorMessageText = "The skill encountered an error or bug.";
                Activity errorMessage =
                    MessageFactory.text(errorMessageText, errorMessageText, InputHints.IGNORING_INPUT);
                return turnContext.sendActivity(errorMessage).thenAccept(result -> {
                    String secondLineMessageText = "To continue to run this bot, please fix the bot source code.";
                    Activity secondErrorMessage =
                        MessageFactory.text(secondLineMessageText, secondLineMessageText, InputHints.EXPECTING_INPUT);
                    turnContext.sendActivity(secondErrorMessage)
                        .thenApply(
                            sendResult -> {
                                // Send a trace activity, which will be displayed in the Bot Framework Emulator.
                                // Note: we return the entire exception in the value property to help the
                                // developer;
                                // this should not be done in production.
                                return TurnContext.traceActivity(
                                    turnContext,
                                    String.format("OnTurnError Trace %s", exception.toString())
                                );

                            }
                        );
                });
            } catch (Exception ex) {
                return Async.completeExceptionally(ex);
            }
        }

        private CompletableFuture<Void> sendEoCToParent(TurnContext turnContext, Throwable exception) {
            try {
                // Send an EndOfConversation activity to the skill caller with the error to end
                // the conversation,
                // and let the caller decide what to do.
                Activity endOfConversation = Activity.createEndOfConversationActivity();
                endOfConversation.setCode(EndOfConversationCodes.SKILL_ERROR);
                endOfConversation.setText(exception.getMessage());
                return turnContext.sendActivity(endOfConversation).thenApply(result -> null);
            } catch (Exception ex) {
                return Async.completeExceptionally(ex);
            }
        }

    }
}
